src/mds/Server.cc

```diff
@@ -6290,6 +6298,142 @@ void Server::handle_client_removexattr(MDRequestRef& mdr)
   journal_and_reply(mdr, cur, 0, le, new C_MDS_inode_update_finish(this, mdr, cur));
 }
 
+void Server::handle_client_getvxattr(MDRequestRef& mdr)
+{
+  const auto& req = mdr->client_request;
+  string name{req->get_path2()};
+
+  // is a ceph virtual xattr?
+  if (!is_ceph_vxattr(name)) {
+    respond_to_request(mdr, -CEPHFS_EINVAL);
+    return;
+  }
+
+  CInode *cur = rdlock_path_pin_ref(mdr, true, false);
+  if (!cur) {
+    return;
+  }
+
+  if (!cur->is_dir()) {
+    respond_to_request(mdr, -CEPHFS_ENOTDIR);
+    return;
+  }
+
+  std::stringstream ss;
+  int r = 0;
+  ceph::bufferlist bl;
+  // handle these vxattrs
+  if (name.substr(0, 15) == "ceph.dir.layout"sv) {
+    std::string layout_field;
+
+    struct layout_xattr_info_t {
+      const file_layout_t  layout;
+      const std::string    inheritance;
+
+      layout_xattr_info_t(const file_layout_t& l, const std::string& inh)
+        : layout(l), inheritance(inh) { }
+    };
+
+    auto is_default_layout = [](const file_layout_t& layout) -> bool {
+      auto default_layout = file_layout_t::get_default();
+      return (layout == default_layout);
+    };
+    auto get_inherited_layout = [&](CInode *cur) -> layout_xattr_info_t {
+      auto orig_in = cur;
+
+      while (cur) {
+        if (cur->get_projected_inode()->has_layout()) {
+	  auto& curr_layout = cur->get_projected_inode()->layout;
+	  if (is_default_layout(curr_layout)) {
+	    return {curr_layout, "default"s};
+	  }
+          if (cur == orig_in) {
+	      // we've found a new layout at this inode
+	      return {curr_layout, "new"s};
+          } else {
+	      return {curr_layout, "inherited"s};
+          }
+        }
+
+        if (cur->is_root()) {
+          break;
+	}
+
+        cur = cur->get_projected_parent_dir()->get_inode();
+      }
+      ceph_assert("something is really messed up with layouts!" == nullptr);
+      auto& root_layout = mdcache->get_root()->get_projected_inode()->layout;
+      if (is_default_layout(root_layout)) {
+	return {mdcache->get_root()->get_projected_inode()->layout, "default"s};
+      } else {
+	return {mdcache->get_root()->get_projected_inode()->layout, "new"s};
+      }
+    };
+
+    const auto lxi = get_inherited_layout(cur);
+
+    if (name == "ceph.dir.layout"sv) {
+      ss << "{\"stripe_unit\": " << lxi.layout.stripe_unit
+	 << ", \"stripe_count\": " << lxi.layout.stripe_count
+	 << ", \"object_size\": " << lxi.layout.object_size
+	 << ", \"pool_name\": ";
+      mds->objecter->with_osdmap([lxi, &ss](const OSDMap& o) {
+	  ss << "\"";
+          if (o.have_pg_pool(lxi.layout.pool_id))
+	    ss << o.get_pool_name(lxi.layout.pool_id);
+	  ss << "\"";
+          });
+      ss << ", \"pool_id\": " << (uint64_t)lxi.layout.pool_id;
+      if (lxi.layout.pool_ns.length())
+	ss << ", \"pool_namespace\": \"" << lxi.layout.pool_ns << "\"";
+
+      ss << ", \"inheritance\": \"@" << lxi.inheritance << "\"}";
```

Comment:
There's a downstream BZ that has a request to make the layout and the layout inheritance status available at any point in the dir structure. This makes it more user friendly, rather than require the user to hunt the layout by walking up the path.